File: 0000-0999/0395.LongestSubstringwithAtLeastKRepeatingCharacters.go
func longestSubstring(s string, k int) int {
    n := len(s)
    if k <= 1 {
        return n
    }
    if n < k {
        return 0
    }
    cnt := make(map[byte]int)
    for i := range s {
        cnt[s[i]]++
    }
    var minCh byte
    minCnt := n+1
    for ch, c := range cnt {
        if c < minCnt {
            minCh, minCnt = ch, c
        }
    }
    if minCnt >= k {
        return n
    }
    subs := strings.Split(s, string(minCh))
    ans := 0
    for _, sub := range subs {
        ans = max(ans, longestSubstring(sub, k))
    }
    return ans
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: 0000-0999/0395.LongestSubstringwithAtLeastKRepeatingCharacters.py
class Solution(object):
    def longestSubstring(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        for ch in set(s):
            if s.count(ch) < k:
                return max(self.longestSubstring(sub, k) for sub in s.split(ch))
        return len(s)
        '''# recursive
        if len(s) < k:
            return 0
        dic = {}
        for ch in s:
            dic[ch] = dic.get(ch, 0) + 1
        index = 0
        while index < len(s) and dic[s[index]] >= k:
            index += 1
        if index == len(s):
            return index
        l = self.longestSubstring(s[:index], k)
        r = self.longestSubstring(s[index+1:], k)
        return max(l, r)
        '''